hudi-client/hudi-client-common/src/test/java/org/apache/hudi/client/transaction/TestTransactionManager.java

```diff
@@ -0,0 +1,208 @@
+
+public class TestTransactionManager extends HoodieCommonTestHarness {
+  HoodieWriteConfig writeConfig;
+  TransactionManager transactionManager;
+
+  @BeforeEach
+  private void init() throws IOException {
+    initPath();
+    initMetaClient();
+    this.writeConfig = getWriteConfig();
+    this.transactionManager = new TransactionManager(this.writeConfig, this.metaClient.getFs());
+  }
+
+  private HoodieWriteConfig getWriteConfig() {
+    return HoodieWriteConfig.newBuilder()
+        .withPath(basePath)
+        .withCompactionConfig(HoodieCompactionConfig.newBuilder()
+            .withFailedWritesCleaningPolicy(HoodieFailedWritesCleaningPolicy.LAZY)
+            .build())
+        .withWriteConcurrencyMode(WriteConcurrencyMode.OPTIMISTIC_CONCURRENCY_CONTROL)
+        .withLockConfig(HoodieLockConfig.newBuilder()
+            .withLockProvider(InProcessLockProvider.class)
+            .build())
+        .build();
+  }
+
+  @Test
+  public void testSingleWriterTransaction() {
+    transactionManager.beginTransaction();
+    transactionManager.endTransaction();
+  }
+
+  @Test
+  public void testSingleWriterNestedTransaction() {
+    transactionManager.beginTransaction();
+    assertThrows(HoodieLockException.class, () -> {
+      transactionManager.beginTransaction();
+    });
+
+    transactionManager.endTransaction();
+    assertThrows(HoodieLockException.class, () -> {
+      transactionManager.endTransaction();
+    });
+  }
+
+  @Test
+  public void testSingleWriterMultipleTransactions() {
+    for (int i = 0; i < 32; i++) {
+      transactionManager.beginTransaction();
+      transactionManager.endTransaction();
+    }
+  }
+
+  @Test
+  public void testMultiWriterTransactions() {
+    final int threadCount = 3;
+    final long awaitMaxTimeoutMs = 2000L;
+    final CountDownLatch latch = new CountDownLatch(threadCount);
+    final AtomicBoolean writer1Completed = new AtomicBoolean(false);
+    final AtomicBoolean writer2Completed = new AtomicBoolean(false);
+
+    // Let writer1 get the lock first, then wait for others
+    // to join the sync up point.
+    Thread writer1 = new Thread(() -> {
+      assertDoesNotThrow(() -> {
+        transactionManager.beginTransaction();
+      });
+      latch.countDown();
+      try {
+        latch.await(awaitMaxTimeoutMs, TimeUnit.MILLISECONDS);
+        // Following sleep is to make sure writer2 attempts
+        // to try lock and to get bocked on the lock which
+        // this thread is currently holding.
+        Thread.sleep(50);
+      } catch (InterruptedException e) {
+        //
+      }
+      assertDoesNotThrow(() -> {
+        transactionManager.endTransaction();
+      });
+      writer1Completed.set(true);
+    });
+    writer1.start();
+
+    // Writer2 will block on trying to acquire the lock
+    // and will eventually get the lock before the timeout.
+    Thread writer2 = new Thread(() -> {
+      latch.countDown();
```

Comment:
We should actually use `CyclicBarrier` instead of Latch here if we want to be sure threads are being interleaved -- since we don't do any memory barriers in b/w initialization of W1 and W2 it's still possible for W2 to complete before W1 takes the lock (since it doesn't wait on the latch

Reply:
Writer1 does the begin transaction first and then does countDown next. Then it sleeps for sometime and then ends the transaction.
Writer2 does the countDown first and then does begin transaction. So, we will have the writer1:begin_txn, writer2:begin_txn, writer1:end_txn, writer2:end_txn. I can make it more stronger by replacing the sleep in Writer1 with one more latch, but its not needed strictly.

Reply:
While i agree that this is improbable -- we can't rule out tho that JLS does not rule out possibility that W2 will be able to grab the lock first (b/c there're no explicit sync barriers between those that will guarantee happens-before relationship)

Reply:
Made both the threads do countDown() and await() which will force the happens-before relationship between these two threads and the make the test deterministic.